text/0000-rust-for-linux-project-goal-2024h2.md

```diff
@@ -0,0 +1,239 @@
+
+### Labeled goto in inline assembler and extended `offset_of!` support
+
+These are two smaller extensions required by the Rust-for-Linux kernel support.
+Both have been implemented but more experience and/or may be needed before stabilization is accepted.
+
+### RFL on Rust CI
+
+> **Update**: This was completed in [PR #125209] by Jakub Beránek during the planning process!
+
+[PR #125209]: https://github.com/rust-lang/rust/pull/125209
+
+Rust sometimes integrates external projects of particular importance or interest into its CI.
+This gives us early notice when changes to the compiler or stdlib impact that project.
+Some of that breakage is accidental, and CI integration ensures we can fix it without the project ever being impacted.
+Otherwise the breakage is intentional, and this gives us an early way to notify the project so they can get ahead of it.
+
+Because of the potential to slow velocity and incur extra work,
+the bar for being integrated into CI is high, but we believe that Rust For Linux meets that bar.
+Given that RFL would not be the first such project to be integrated into CI,
+part of pursuing this goal should be establishing clearer policies on when and how we integrate external projects into our CI,
+as we now have enough examples to generalize somewhat.
+
+### Pointers to statics in constants
+
+The RFL project has a need to create vtables in read-only memory (unique address not required). The current implementation relies on the `const_mut_refs` and `const_refs_to_static` features ([representative example](https://godbolt.org/z/r58jP6YM4)). Discussion has identified some questions that need to be resolved but no major blockers.
+
+## The "shiny future" we are working towards
+
+The ultimate goal is to enable smooth and ergonomic interop between Rust and the Linux kernel's idiomatic data structures.
+
+In addition to the work listed above, there are a few other obvious items that the Rust For Linux project needs. If we can find owners for these this year, we could even get them done as a "stretch goal":
+
+### Custom builds of core/alloc with specialized configuration options
+
+The RFL project builds the stdlib with a number of configuration options to eliminate undesired aspects of libcore (listed in [RFL#2][]). They need a standard way to build a custom version of core as well as agreement on the options that the kernel will continue using.
+
+### Stable sanitizer support
+
+Support for building and using sanitizers, in particular KASAN.
+
+### Code-generation features and compiler options
+
+The RFL project requires various code-generation options. Some of these are related to custom features of the kernel, such as [X18 support][#748] but others are codegen options like sanitizers and the like. Some subset of the options listed on [RFL#2][] will need to be stabilized to support being built with all required configurations, but working out the precise set will require more effort.
```

Comment:
A number of these options (for example the reserve `x18` option) need to be set when building the parts of the standard library that RfL uses. If the "stabilize options for building core/alloc with fewer features" step is achieved, that provides a natural way to configure those options. 

If the step is not achieved, do we have plan on record for making RfL able to use these codegen options? It might be good to at least have a fallback option identified to maximize the value delivered from the rest of the work (for example, we could commit to shipping a set of `*-linux-kernel` targets at Tier 2 or above that have the appropriate codegen options set).

Reply:
> that provides a natural way to configure those options.

Do you mean they would be inherited for the rest of the code so they would not need to be passed to other `rustc` calls? Or am I misunderstanding?

In general, I feel that the less special `core` and `alloc` are, and the more flexibility on codegen there is, the better it would be for the kernel and similar projects.

So being able to compile `core` and `alloc` just (almost) like any other crate (like we do now), and being able to customize it as much as possible (as much as C compilers allow) with simple CLI flags on `rustc`, is ideal.

Having fixed codegen options in built-in targets that can only be tweaked on the Rust side sounds too rigid, and we would also have the issue of the potentially many combinations of targets.

Reply:
No, what I mean is that if we have a stable option to allow the user to build standard library crates, I would think that would naturally include the ability for a user to specify additional codegen flags during that compilation. Thus, "how does RfL enable things like KASAN, reserve `x18`, etc?" is resolved. 

If we do not have a stable way for users to build the standard library crates, we should consider providing some other way for RfL to get the standard library crates you need without use of nightly features to build them.

Reply:
> Thus, "how does RfL enable things like KCFI, reserve `x18`, etc?" is resolved.

But why would it be resolved? We would still need to pass the flags to other translation units / crates. (That is why I was asking whether you meant that somehow those flags got inherited if they depend on a `core` that was compiled with them.)

Reply:
If you can you build the standard library (or parts there of) on a stable compiler, I'm assuming you can just pass along `-Creserve-x18` or whatever you need 🙂 

My broader point though is that if we've done the work to enable you to move off nightly features for these other things, we should also provide _some_ path for you to use `libcore`, etc with the required codegen options in a stable way as well.

Reply:
Ah, if the options are stable already, of course! (and, for us, just using the flags normally like we would for any other TU is ideal). I thought you meant that, *without* having stabilized those flags, somehow solving the issue for building the standard library with them in a stable way would provide a solution for the rest of the code (and thus the inheriting question).

Agreed that, if the codegen flags are stabilized, then we need a stable way to build `core`/`alloc` with those too. (The text in the other section could have explicitly mentioned the flags too to be more clear, i.e. one could read that we only need to use `cfg`s and not the codegen options there, I guess)

Thanks for clarifying!